File: lakecontrol.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  main.py
#  
#  
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#  
#  
#import pyvisa as visa
import sys  # sys нужен для передачи argv в QApplication
from PyQt5 import QtCore, QtWidgets
from PyQt5.QtCore import pyqtSignal, pyqtSlot
import tcontrol_ui
import pyvisa
# import threading
# from concurrent.futures import ThreadPoolExecutor
# import numpy as np
from time import localtime, strftime, time
import logging

# class TestApp(QtWidgets.QWidget, visalab_dialogs.sweep_Ui_Form):
	# def __init__(self):
		# super().__init__()
		# self.setupUi(self)
		
format = "%(asctime)s: %(message)s"
logging.basicConfig(format=format, 
	level=logging.INFO, 
	datefmt="%H:%M:%S")
# comment next line if debug is not needed	
# logging.getLogger().setLevel(logging.DEBUG)


class lakecontrolApp(QtWidgets.QDialog, tcontrol_ui.Ui_Dialog):
	
	
	def __init__(self):
		super().__init__()
		self.setupUi(self)
		
		#self.rm = pyvisa.ResourceManager()
		self.state={}
		self.current_dev = None
		self.analogCombo.addItems(['1','2'])
		self.loopCombo.addItems(['1'])
		self.hrangeCombo.addItems(['0', '1', '2', '3', '4', '5'])
		self.programCombo.addItems(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10'])							
		
		self.prunBtn.setEnabled(False)
		self.ptermBtn.setEnabled(False)
		self.hoffBtn.setEnabled(False)
		self.rampSwitch.setEnabled(False)
		
		
		self.getButton.clicked.connect(self.onGet)
		self.setpUpd.clicked.connect(self.onSetp)
		self.rampUpd.clicked.connect(self.onRamp)
		self.heaterUpd.clicked.connect(self.onHupd)
		self.hoffBtn.clicked.connect(self.onHoff)
		self.scanButton.clicked.connect(self.onScan)
		self.rampSwitch.stateChanged.connect(self.rSwitchchange)
		self.powerSwitch.stateChanged.connect(self.pSwitchchange)
		self.setResBtn.clicked.connect(self.onSetRes)
		self.exitButton.clicked.connect(self.close)
		self.analogUpd.clicked.connect(self.onAUpd)
		self.instrList.currentTextChanged.connect(self.onChange)
		self.prunBtn.clicked.connect(self.onRun)		
		self.ptermBtn.clicked.connect(self.onTerminate)		
		
	def onRun(self):
		if not self.current_dev == None:
			self.current_dev.write(f'PGMRUN {self.programCombo.currentText()}')
			self.prunBtn.setEnabled(False)
			self.ptermBtn.setEnabled(True)
			self.messageBox.append(f'Program {self.programCombo.currentText()} started')
	
	def onTerminate(self):
		if not self.current_dev == None:
			self.current_dev.write(f'PGMRUN 0')
			self.prunBtn.setEnabled(True)
			self.ptermBtn.setEnabled(False)
			self.messageBox.append(f'Program terminated')
	
	def onChange(self):
		self.current_dev = self.rm.open_resource(self.instrList.currentText())
		logging.info(f'onChange: current device is {self.instrList.currentText()}\n')
		self.updateState()
	
	def updateState(self):
		l=''
		st = self.current_dev.query('AOUT?1').strip()
		l += 'AnalogOut1: ' + st + '%\n'
		st = self.current_dev.query('AOUT?2').strip()
		l += 'AnalogOut2: ' + st + '%\n'
		st = self.current_dev.query('HTR?').strip()
		l += 'Heater output: ' + st + '%\n'
		self.houtSpin.setValue(float(st))
		st = self.current_dev.query('RAMP?1').strip()
		l += f'Ramp state: {st}\n'
		try:
			ramp_on, ramp_rate = st.split(',')
			if ramp_on == '1':
				ramp = 'ON'
				self.rampSwitch.setChecked(True)
			elif ramp_on == '0':
				ramp = 'OFF'
				self.rampSwitch.setChecked(False)
			l += 'Ramp settings: ' + f'Ramp {ramp}, Ramp rate={ramp_rate}K/min' + '\n'
			self.rampSpin.setValue(float(ramp_rate))
		except Exception:
			logging.info('Unable to unpack ramp setings')
		
		st = self.current_dev.query('RAMPST?1').strip()
		if st == '0':
			st = 'idle'
		if st == '1':
			st = 'running'
		l += 'Ramp status: ' + st + '\n'
		st = self.current_dev.query('SETP?1').strip()
		l += 'Current setpoint: ' + st + 'K\n'	
		self.setpSpin.setValue(float(st))	
		st = self.current_dev.query('RANGE?').strip()
		if st == '0':
			l += 'Heater off'+'\n'
			self.hoffBtn.setEnabled(False)
		else:
			l += 'Heater range: ' + st + '\n'
			self.hoffBtn.setEnabled(True)
			
		self.hrangeCombo.setCurrentIndex(int(st))
		st = self.current_dev.query('CDISP?1').strip()
		disp_list = st.split(',')
		self.hResSpin.setValue(float(disp_list[1]))
		l += 'Display settings: ' + st + '\n'	
		st = self.current_dev.query('CSET?1').strip()
		loop_list = st.split(',')
		if loop_list[-1] == '1':
			self.powerSwitch.setChecked(True)
		elif loop_list[-1] == '0': 
			self.powerSwitch.setChecked(False)
		l += 'Loop1 settings: ' + st + '\n'
		st = self.current_dev.query('PGMRUN?').strip()
		program, status = st.split(',')
		if program == '00':
			self.prunBtn.setEnabled(True)
			self.ptermBtn.setEnabled(False)
		else:
			l +=f'Program {program} is running'
			self.prunBtn.setEnabled(False)
			self.ptermBtn.setEnabled(True)
		self.ConfigBox.setText(l)
				
	
	def onAUpd(self):
		logging.debug('onAUpd: Update AnalogOut clicked!')
		if not self.current_dev == None:
			self.current_dev.write(f'ANALOG {self.analogCombo.currentText()},0,2,,,,,{self.analogSpin.value()}')
			self.messageBox.append(f'Analog out {self.analogCombo.currentText()} is set to {self.analogSpin.value()}')
	
	
	def onSetRes(self):
		logging.debug('onSetRes: Set resistivity clicked!')
		self.messageBox.append('Set resistivity clicked!')
		self.messageBox.append(f'Heater resistance is set to {self.hResSpin.value()}')
	
	def rSwitchchange(self, state):
		logging.info(f'Ramp switch state: {state}!')
		if state == 2 and not self.current_dev == None:
			self.current_dev.write(f'RAMP 1, 1, {self.rampSpin.value()}')
			self.messageBox.append('Ramp on')
		elif state == 0 and not self.current_dev == None:
			self.current_dev.write(f'RAMP 1, 0')
			self.messageBox.append('Ramp off')
			
	
	def pSwitchchange(self, state):
		logging.debug(f'Power switch state: {state}!')
		if state == 2 and not self.current_dev == None:
			self.current_dev.write(f'CSET 1,,,,1')
			self.messageBox.append('Loop1 powerup on')
		elif state == 0 and not self.current_dev == None:
			self.current_dev.write(f'CSET 1,,,,0')
			self.messageBox.append('Loop1 powerup off')
	
	def onGet(self):
		logging.debug('onGet: Get clicked!')
		if not self.current_dev == None:
			self.updateState()
	
	def onSetp(self):
		logging.debug('onSetp: Update setpoint clicked!')
		if not self.current_dev == None:
			self.current_dev.write(f'SETP 1,{self.setpSpin.value()}')
			self.messageBox.append(f'Setpoint is set to {self.setpSpin.value()}K')
		
	def onRamp(self):
		logging.debug('onRamp: Update ramp clicked!')
		if not self.current_dev == None:
			self.current_dev.write(f'RAMP 1,1,{self.rampSpin.value()}')
			self.messageBox.append(f'Ramp is set to {self.rampSpin.value()}K/min')
		
	def onHupd(self):
		logging.info('onHupd: Update heater clicked!')
		if not self.current_dev == None:
			self.current_dev.write(f'RANGE {self.hrangeCombo.currentText()}')
			self.messageBox.append(f'Heater range changed to {self.hrangeCombo.currentText()}')
			self.hoffBtn.setEnabled(True)
		
	def onHoff(self):
		logging.info('Heater off clicked!')
		if not self.current_dev == None:
			self.current_dev.write('RANGE 0')
			self.hoffBtn.setEnabled(False)
			self.messageBox.append('Heater off')
		
	def onScan(self):
		#logging.info('Scan clicked!')
		self.instrList.clear()
		self.rm = pyvisa.ResourceManager()
		self.dev_list = self.rm.list_resources()
		for dev in self.dev_list:
			if not dev.find('GPIB') == -1:
				try:
					inst = self.rm.open_resource(dev)
					id_str = inst.query("*IDN?")
					if not id_str.find('MODEL340')==-1:
						self.instrList.addItem(dev)
						logging.info('onScan: LS340 found!')
				except Exception:
					logging.exception(f'OnScan: Error communicating with device {dev}!')
		if len(self.instrList)>0:
			self.current_dev = self.rm.open_resource(self.instrList.currentText())
			self.updateState()
			logging.info(f'onScan: current device is {self.instrList.currentText()}\n')
			self.prunBtn.setEnabled(True)
			self.ptermBtn.setEnabled(True)
			self.hoffBtn.setEnabled(True)
			self.rampSwitch.setEnabled(True)
		else:
			logging.info(f'Lakeshore340 devices on GPIB bus not found')
		
		
		
					
def main(args):
	app = QtWidgets.QApplication(sys.argv)  # Новый экземпляр QApplication
	window = lakecontrolApp()  # Создаём объект класса visaLabApp
	window.show()  # Показываем окно
	app.exec_()  # и запускаем приложение

if __name__ == '__main__':
	import sys
	sys.exit(main(sys.argv))

File: lakecontrol.cpp
#include "lakecontrol.h"
#include "ni488.h"
#include "windows.h"

LakeControl::LakeControl(QWidget *parent) : QWidget(parent)
{
    setupUi(this);
    connect(exitButton, SIGNAL(clicked()),this, SLOT(close()));
    connect(getButton,SIGNAL(clicked()),this,SLOT(onGetState()));
    connect(scanButton,SIGNAL(clicked()),this,SLOT(onScan()));
    connect(pointButton,SIGNAL(clicked()),this,SLOT(onSetpoint()));
    connect(hoffButton,SIGNAL(clicked()),this,SLOT(onHeaterOff()));
    connect(heaterButton,SIGNAL(clicked()),this,SLOT(onSetRange()));
    connect(resButton,SIGNAL(clicked()),this,SLOT(onSetResistance()));
    connect(rateButton,SIGNAL(clicked()),this,SLOT(onRamp()));
    connect(rampCheck,SIGNAL(stateChanged(int)),this,SLOT(onRampCheck(int)));
    connect(powerCheck,SIGNAL(stateChanged(int)),this,SLOT(onPowerCheck(int)));
    connect(enCheck,SIGNAL(stateChanged(int)),this,SLOT(onEnableCheck(int)));
    connect(instrList,SIGNAL(currentIndexChanged(int)),this,SLOT(onGetState()));
    gboard board();
    pointButton->setEnabled(false);
    hoffButton->setEnabled(false);
    heaterButton->setEnabled(false);
    resButton->setEnabled(false);
    rateButton->setEnabled(false);
    rampCheck->setEnabled(false);
    powerCheck->setEnabled(false);
    }

void LakeControl::onGetState()
{
// ANALOG?<output> - <bipolar enable>,<mode>,<input>,<source>,<high value>,
    //<low value>,<manual value>, format: n,n,ann,n,+-nnn.nnnE+-n,+-nnn.nnnE+-n,+-nnn.n[term]
//HTR? - heater output - nnn.nn[term]
//RAMP?<loop> - ramp settings - <off/on>,<rate value>, format:nnn.n[term]
//RAMPST?<loop> - ramp status, format:n[term]
//SETP?<loop> - current setpoint, format: +-nnn.nnE+-n[term]
//RANGE? - heater range, format: n[term]
//CDISP?
    //CLIMI?
    if(instrList->currentText()==NULL)
    {
        stateEdit->setText("No instrument selected!");
        return;
    }
    gdevice dev(comboBox->currentText().toInt(),extractAdress(instrList->currentText()),
                extractType(instrList->currentText()));
//    board.writeCommand(dev, "HTRST?");
//    QString data=board.receiveData(dev);
    stateEdit->clear();
//    if(data=="05")
//    {
//        stateEdit->append("Open Heater Load!!! "+data);
//    }
//    if(data=="06")
//    {
//        stateEdit->append("Heater Load is less than 10 Ohm!!! "+data);
//    }
    board.writeCommand(dev, "ANALOG?1");
    QString data=board.receiveData(dev);
    stateEdit->append("Analog output 1: "+data);
    board.writeCommand(dev, "ANALOG?2");
    data=board.receiveData(dev);
    stateEdit->append("Analog output 2: "+data);
    board.writeCommand(dev, "AOUT?1");
    data=board.receiveData(dev);
    stateEdit->append("Analog output 1 value: "+data);
    board.writeCommand(dev, "AOUT?2");
    data=board.receiveData(dev);
    stateEdit->append("Analog output 2 value: "+data);
    board.writeCommand(dev, "HTR?");
    data=board.receiveData(dev);
    stateEdit->append("Heater output: "+data);
    board.writeCommand(dev, "RAMP?1");
    data=board.receiveData(dev);
    if(data.section(',',0,0).toShort())
            rampCheck->setChecked(true);
    else
        rampCheck->setChecked(false);
    rampSpin->setValue(data.section(',',1,1).toDouble());
    stateEdit->append("Ramp settings: "+data);
    board.writeCommand(dev, "RAMPST?1");
    data=board.receiveData(dev);
    stateEdit->append("Ramp status: "+data);
    board.writeCommand(dev, "CSET?1");
    data=board.receiveData(dev);
    if(data.section(',',3,3).toShort())
            powerCheck->setChecked(true);
    else
        powerCheck->setChecked(false);
    if(data.section(',',2,2).toShort())
            enCheck->setChecked(true);
    else
        enCheck->setChecked(false);
    board.writeCommand(dev, "SETP?1");
    data=board.receiveData(dev);
    stateEdit->append("Current setpoint: "+data);
    pointSpin->setValue(data.toDouble());
    board.writeCommand(dev, "RANGE?");
    data=board.receiveData(dev);
    stateEdit->append("Heater range: "+data);
    if(data.toShort())
       hoffButton->setEnabled(true);
    else
        hoffButton->setEnabled(false);
    board.writeCommand(dev, "CSET?1");
    data=board.receiveData(dev);
    stateEdit->append("Loop 1 settings: "+data);
    board.writeCommand(dev, "CDISP?1");
    data=board.receiveData(dev);
    stateEdit->append("Heater resistance: "+data.section(',',1,1));
    hSpin->setValue(data.section(',',1,1).toDouble());
    board.writeCommand(dev, "CLIMI?");
    data=board.receiveData(dev);
    double range=data.toDouble();
    range=range*range*hSpin->value();
    rangeList->clear();
    rangeList->addItem(QString("5:%1W").arg(range));
    rangeList->addItem(QString("4:%1W").arg(range/10));
    rangeList->addItem(QString("3:%1mW").arg(range*10));
    rangeList->addItem(QString("2:%1mW").arg(range));
    rangeList->addItem(QString("1:%1mW").arg(range/10));
    pointButton->setEnabled(true);
    heaterButton->setEnabled(true);
    resButton->setEnabled(true);
    rateButton->setEnabled(true);
    rampCheck->setEnabled(true);
    powerCheck->setEnabled(true);
}
void LakeControl::onRamp()
{
    if(instrList->currentText()==NULL)
    {
        stateEdit->setText("No instrument selected!\n");
        return;
    }
    gdevice dev(comboBox->currentText().toInt(),extractAdress(instrList->currentText()),
                extractType(instrList->currentText()));
    board.writeCommand(dev, QString("RAMP 1,,%1").arg(rampSpin->value()));
    mBox->append(QString("%1::Ramp rate is set to %2K/min").arg(dev.getAdress()).arg(rampSpin->value()));
}
void LakeControl::onRampCheck(int state)
{
    gdevice dev(comboBox->currentText().toInt(),extractAdress(instrList->currentText()),
               extractType(instrList->currentText()));
    if(state)
    {
        if(instrList->currentText()==NULL)
        {
            stateEdit->setText("No instrument selected!\n");
            return;
        }
        board.writeCommand(dev, "RAMP 1,1");
        mBox->append(QString("%1::Ramp is ON").arg(dev.getAdress()));
    }
    else
    {
        board.writeCommand(dev, "RAMP 1,0");
        mBox->append(QString("%1::Ramp is OFF").arg(dev.getAdress()));
    }
}
void LakeControl::onEnableCheck(int state)
{
    gdevice dev(comboBox->currentText().toInt(),extractAdress(instrList->currentText()),
               extractType(instrList->currentText()));
    if(state)
    {
        if(instrList->currentText()==NULL)
        {
            stateEdit->setText("No instrument selected!\n");
            return;
        }
        board.writeCommand(dev, "CSET 1,,,1");
        mBox->append(QString("%1::Loop 1 control is enabled").arg(dev.getAdress()));
    }
    else
    {
        board.writeCommand(dev, "CSET 1,,,0");
        mBox->append(QString("%1::Loop 1 control is disabled").arg(dev.getAdress()));
    }
}
void LakeControl::onPowerCheck(int state)
{
    gdevice dev(comboBox->currentText().toInt(),extractAdress(instrList->currentText()),
               extractType(instrList->currentText()));
    if(state)
    {
        if(instrList->currentText()==NULL)
        {
            stateEdit->setText("No instrument selected!\n");
            return;
        }
        board.writeCommand(dev, "CSET 1,,,,1");
        mBox->append(QString("%1::Loop 1 powerup is ON").arg(dev.getAdress()));
    }
    else
    {
        board.writeCommand(dev, "CSET 1,,,,0");
        mBox->append(QString("%1::Loop 1 powerup is OFF").arg(dev.getAdress()));
    }
}
void LakeControl::onScan()
{
    board.setId(comboBox->currentText().toInt());
    devList=board.scanForInstruments();
    for(int i=0;i<devList.count();++i)
    {
        if(devList[i].getType().contains("MODEL340"))
        {
            instrList->addItem(QString("%1").arg(devList[i].getAdress())+": "+devList[i].getType());
        }
    }
}
void LakeControl::onSetpoint()
{
    if(instrList->currentText()==NULL)
    {
        stateEdit->setText("No instrument selected!");
        return;
    }
    gdevice dev(comboBox->currentText().toInt(),extractAdress(instrList->currentText()),
                extractType(instrList->currentText()));
    board.writeCommand(dev, QString("SETP 1,%1").arg(pointSpin->value()));
    mBox->append(QString("%1::Setpoint is set to %2").arg(dev.getAdress()).arg(pointSpin->value()));
}
void LakeControl::onHeaterOff()
{
    if(instrList->currentText()==NULL)
    {
        stateEdit->setText("No instrument selected!\n");
        return;
    }
    gdevice dev(comboBox->currentText().toInt(),extractAdress(instrList->currentText()),
                extractType(instrList->currentText()));
    board.writeCommand(dev, "RANGE 0");
    mBox->append(QString("%1::Heater is OFF").arg(dev.getAdress()));
    hoffButton->setEnabled(false);
}
void LakeControl::onSetRange()
{
    if(instrList->currentText()==NULL)
    {
        stateEdit->setText("No instrument selected!\n");
        return;
    }
    gdevice dev(comboBox->currentText().toInt(),extractAdress(instrList->currentText()),
                extractType(instrList->currentText()));
    board.writeCommand(dev, QString("RANGE %1").arg(5-rangeList->currentIndex()));
    mBox->append(QString("%1::Heater range is set to ").arg(dev.getAdress())+rangeList->currentText());
    onGetState();
}
void LakeControl::onSetResistance()
{
    if(instrList->currentText()==NULL)
    {
        stateEdit->setText("No instrument selected!\n");
        return;
    }
    gdevice dev(comboBox->currentText().toInt(),extractAdress(instrList->currentText()),
                extractType(instrList->currentText()));
    board.writeCommand(dev, QString("CDISP 1,,%1").arg(hSpin->value()));
    mBox->append(QString("%1::Heater resistance is set to %2 Ohm").arg(dev.getAdress()).arg(hSpin->value()));
}
int LakeControl::extractAdress(QString line)
{
    int first=line.indexOf(":");
    return line.left(first).toInt();

}
QString LakeControl::extractType(QString line)
{
    int first=line.indexOf(":");
    return line.right(line.size()-first-1);
}



